staging/src/k8s.io/apiserver/pkg/util/peerproxy/peerproxy_handler.go

```diff
@@ -135,111 +116,171 @@ func (h *peerProxyHandler) WrapHandler(handler http.Handler) http.Handler {
+		// find servers that are capable of serving this request
+		peerEndpoint, err := h.findServiceableByServers(r.Context(), gvr, apiserversi)
+		if err != nil {
+			klog.ErrorS(err, "error fetching remote server details while proxying")
+			responsewriters.ErrorNegotiated(apierrors.NewServiceUnavailable("Error getting ip and port info of the remote server while proxying"), h.serializer, gvr.GroupVersion(), w, r)
 			return
 		}
 
-		// no apiservers were found that could serve the request, pass request to
-		// next handler, that should eventually serve 404
-
-		// TODO: maintain locally serviceable GVRs somewhere so that we dont have to
-		// consult the storageversion-informed map for those
-		if len(serviceableByResp.peerEndpoints) == 0 {
-			klog.Error(fmt.Sprintf("GVR %v is not served by anything in this cluster", gvr))
-			handler.ServeHTTP(w, r)
-			return
+		if peerEndpoint != "" {
+			h.proxyRequestToDestinationAPIServer(r, w, peerEndpoint)
 		}
+	})
+}
 
-		// otherwise, randomly select an apiserver and proxy request to it
-		rand := rand.Intn(len(serviceableByResp.peerEndpoints))
-		destServerHostPort := serviceableByResp.peerEndpoints[rand]
-		h.proxyRequestToDestinationAPIServer(r, w, destServerHostPort)
+func (h *peerProxyHandler) apiserverIdentityLeases() ([]*v1.Lease, error) {
+	var apiserversi []*v1.Lease
+	apiserversi, err := h.apiserverIdentityLister.Leases(metav1.NamespaceSystem).List(labels.SelectorFromSet(labels.Set{"apiserver.kubernetes.io/identity": "kube-apiserver"}))
+	if err != nil {
+		return apiserversi, err
+	}
 
-	})
+	return apiserversi, nil
 }
 
-func (h *peerProxyHandler) findServiceableByServers(gvr schema.GroupVersionResource, localAPIServerId string, reconciler reconcilers.PeerEndpointLeaseReconciler) (serviceableByResponse, error) {
+func (h *peerProxyHandler) shouldServeLocally(apiserversi []*v1.Lease, gvr schema.GroupVersionResource) bool {
+	if len(apiserversi) <= 1 {
+		return true
+	}
 
-	apiserversi, ok := h.svMap.Load(gvr)
+	return h.localDiscoverySuccessful(gvr)
+}
 
-	// no value found for the requested gvr in svMap
-	if !ok || apiserversi == nil {
-		return serviceableByResponse{}, fmt.Errorf("no StorageVersions found for the GVR: %v", gvr)
+func (h *peerProxyHandler) localDiscoverySuccessful(gvr schema.GroupVersionResource) bool {
+	discoveryClient, err := discovery.NewDiscoveryClientForConfig(h.loopbackClientConfig)
+	if err != nil {
+		klog.Errorf("error creating discovery client: %v\n", err)
+		return false
 	}
-	apiservers := apiserversi.(*sync.Map)
-	response := serviceableByResponse{}
-	var peerServerEndpoints []string
-	apiservers.Range(func(key, value interface{}) bool {
-		apiserverKey := key.(string)
-		if apiserverKey == localAPIServerId {
-			response.errorFetchingAddressFromLease = true
-			response.locallyServiceable = true
-			// stop iteration
-			return false
+
+	_, resourcesByGV, _, err := discoveryClient.GroupsAndMaybeResources()
```

Comment:
Since we were discussing optimizations, we don't even need uuid/other checks for this right? A simple request on startup + save to map + fetch from map would suffice and only incur 2 requests on apiserver startup with 0 additional requests being sent per resource request.